keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,540 @@
+  endpoint:
+    bucketHost: foo.bar.com
+    bucketPort: 8080
+    bucketName: my-photos-1xj4a
+    region: # provisioner dependent
+    subRegion: # provisioner dependent
+    additionalConfigData: [] #string:string
+  additionalState: [] #string:string
+status:
+  phase: {"Bound", "Released", "Failed"} [8]
+```
+1. name is constructed in the pattern: obc-OBC_NAMESPACE-OBC_NAME
+1. namespace of the object store service as defined in the storage class
+1. the label value shown is the name of the provisioner but due to Kubernetes restrictions slash (/) is replaced by a dash (-).
+1. finalizers set and cleared by the lib's OBC controller. Prevents accidental deletion of an OB.
+   replaced by a dash (-). In this example the provisioner name is `aws-s3.io/bucket`.
+1. name of the storage class, referenced by the OBC, containing the provisioner and object store service name.
+1. objectReference to the associated OBC.
+1. reclaim policy from the Storge Class referenced in the OBC.
+1. phase is the current state of the ObjectBucket:
+    - _Bound_: the operator finished processing the request and linked the OBC and OB
+    - _Released_: the OBC has been deleted, leaving the OB unclaimed.
+    - _Failed_: not currently set.
+
+#### Secret (created by library)
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: MY-BUCKET-1 [1]
+  namespace: OBC-NAMESPACE [2]
+  finalizers: [3]
+  - objectbucket.io/finalizer
+  labels: [4]
+    bucket-provisioner: aws-s3.io-bucket [5]
+  ownerReferences:
+  - name: MY-BUCKET-1 [6]
+    ...
+type: Opaque
+data:
+  ACCESS_KEY_ID: BASE64_ENCODED-1
+  SECRET_ACCESS_KEY: BASE64_ENCODED-2
+  ... [5]
+```
+1. same name as the OBC. Unique since the secret is in the same namespace as the OBC.
+1. namespce of the originating OBC.
+1. finalizers set and cleared by the lib's OBC controller. Prevents accidental deletion of the Secret.
+1. the library adds a label (seen here) but each provisioner can supply their own labels.
+1. the label value shown is the name of the provisioner but due to Kubernetes restrictions slash (/) is replaced by a dash (-).
+1. ownerReference makes this secret a child of the originating OBC for clean up purposes.
+1. ACCESS_KEY_ID and SECRET_ACCESS_KEY are the only secret keys defined by the library.
+Provisioners are able to cause the lib to create additional keys by returning  the `AdditionalSecretConfig` field.
+**Note:** the library will create the Secret using `stringData:` and let the Secret API base64 encode the values.
+
+#### ConfigMap (created by library)
+
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: MY-BUCKET-1 [1]
+  namespace: OBC-NAMESPACE [2]
+  finalizers: [3]
+  - objectbucket.io/finalizer
+  labels: [4]
+    bucket-provisioner: aws-s3.io-bucket [5]
+  ownerReferences: [6]
+  - name: MY-BUCKET-1
+    ...
+data: 
+  BUCKET_HOST: http://MY-STORE-URL [7]
+  BUCKET_PORT: 80 [8]
+  BUCKET_NAME: MY-BUCKET-1 [9]
+  BUCKET_REGION: us-west-1
+  ... [10]
```

Comment:
Pre previous comment - these should be whatever is appropriate for the given data path protocol.

Reply:
If we do this and have a "DataPathProtocol" type field in the OBC then app pods have little chance of being portable across different object store, right? Some degree (obviously not 10%) of portability is/was desired in the original design.

Reply:
The app would be tethered to the protocol regardless as it's expected they've been written using some SDK.   This doesn't restrict apps to a specific object provider, but they do need to be aware of what protocol is being used at the endpoint.  That is, S3 is pretty common, but not ubiquitous, and this must be communicated.